File: app/src/main/java/me/anon/model/Water.java
package me.anon.model;

import android.content.Context;
import android.preference.PreferenceManager;

import java.util.ArrayList;
import java.util.List;

import me.anon.lib.TempUnit;
import me.anon.lib.Unit;

import static me.anon.lib.TempUnit.CELCIUS;
import static me.anon.lib.Unit.ML;

/**
 * // TODO: Add class description
 *
 * @author 7LPdWcaW
 * @documentation // TODO Reference flow doc
 * @project GrowTracker
 */
public class Water extends Action
{
	private Double ppm;
	private Double ph;
	private Double runoff;
	private Double amount;
	private Double temp;
	private List<Additive> additives = new ArrayList<>();

	@Deprecated private Nutrient nutrient;
	@Deprecated private Double mlpl;

	public void setNutrient(Nutrient nutrient)
	{
		this.nutrient = nutrient;
	}

	public void setMlpl(Double mlpl)
	{
		this.mlpl = mlpl;
	}

	public Nutrient getNutrient()
	{
		return nutrient;
	}

	public Double getMlpl()
	{
		return mlpl;
	}

	public Double getPpm()
	{
		return ppm;
	}

	public void setPpm(Double ppm)
	{
		this.ppm = ppm;
	}

	public Double getPh()
	{
		return ph;
	}

	public void setPh(Double ph)
	{
		this.ph = ph;
	}

	public Double getRunoff()
	{
		return runoff;
	}

	public void setRunoff(Double runoff)
	{
		this.runoff = runoff;
	}

	public Double getAmount()
	{
		return amount;
	}

	public void setAmount(Double amount)
	{
		this.amount = amount;
	}

	public Double getTemp()
	{
		return temp;
	}

	public void setTemp(Double temp)
	{
		this.temp = temp;
	}

	public List<Additive> getAdditives()
	{
		return additives;
	}

	public void setAdditives(List<Additive> additives)
	{
		this.additives = additives;
	}

	public String getSummary(Context context)
	{
		Unit measureUnit = Unit.getSelectedMeasurementUnit(context);
		Unit deliveryUnit = Unit.getSelectedDeliveryUnit(context);
		TempUnit tempUnit = TempUnit.getSelectedTemperatureUnit(context);
		boolean usingEc = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("tds_ec", false);

		String summary = "";
		StringBuilder waterStr = new StringBuilder();

		if (getPh() != null)
		{
			waterStr.append("<b>In pH: </b>");
			waterStr.append(getPh());
			waterStr.append(", ");
		}

		if (getRunoff() != null)
		{
			waterStr.append("<b>Out pH: </b>");
			waterStr.append(getRunoff());
			waterStr.append(", ");
		}

		summary += waterStr.toString().length() > 0 ? waterStr.toString().substring(0, waterStr.length() - 2) + "<br/>" : "";

		waterStr = new StringBuilder();

		if (getPpm() != null)
		{
			String ppm = String.valueOf(getPpm().longValue());
			if (usingEc)
			{
				waterStr.append("<b>EC: </b>");
				ppm = String.valueOf((getPpm() * 2d) / 1000d);
			}
			else
			{
				waterStr.append("<b>PPM: </b>");
			}

			waterStr.append(ppm);
			waterStr.append(", ");
		}

		if (getAmount() != null)
		{
			waterStr.append("<b>Amount: </b>");
			waterStr.append(ML.to(deliveryUnit, getAmount()));
			waterStr.append(deliveryUnit.getLabel());
			waterStr.append(", ");
		}

		if (getTemp() != null)
		{
			waterStr.append("<b>Temp: </b>");
			waterStr.append(CELCIUS.to(tempUnit, getTemp()));
			waterStr.append("º").append(tempUnit.getLabel()).append(", ");
		}

		summary += waterStr.toString().length() > 0 ? waterStr.toString().substring(0, waterStr.length() - 2) + "<br/>" : "";

		waterStr = new StringBuilder();

		if (getAdditives().size() > 0)
		{
			waterStr.append("<b>Additives:</b>");

			for (Additive additive : getAdditives())
			{
				if (additive == null || additive.getAmount() == null) continue;

				double converted = ML.to(measureUnit, additive.getAmount());
				String amountStr = converted == Math.floor(converted) ? String.valueOf((int)converted) : String.valueOf(converted);

				waterStr.append("<br/>&nbsp;&nbsp;&nbsp;&nbsp;• ");
				waterStr.append(additive.getDescription());
				waterStr.append("  -  ");
				waterStr.append(amountStr);
				waterStr.append(measureUnit.getLabel());
				waterStr.append("/");
				waterStr.append(deliveryUnit.getLabel());
			}
		}

		summary += waterStr.toString();

		return summary;
	}
}

File: app/src/main/java/me/anon/model/Water.kt
package me.anon.model

import android.content.Context
import android.os.Parcelable
import android.preference.PreferenceManager
import com.squareup.moshi.JsonClass
import kotlinx.android.parcel.Parcelize
import me.anon.grow.R
import me.anon.lib.TempUnit
import me.anon.lib.Unit

/**
 * // TODO: Add class description
 */
@Parcelize
@JsonClass(generateAdapter = true)
class Water(
	var ppm: Double? = null,
	var ph: Double? = null,
	var runoff: Double? = null,
	var amount: Double? = null,
	var temp: Double? = null,
	var additives: ArrayList<Additive> = arrayListOf(),

	override var date: Long = System.currentTimeMillis(),
	override var notes: String? = null
) : Action(date, notes), Parcelable
{
	@Deprecated("")
	public var nutrient: Nutrient? = null
	@Deprecated("")
	public var mlpl: Double? = null

	public fun getSummary(context: Context): String
	{
		val measureUnit = Unit.getSelectedMeasurementUnit(context)
		val deliveryUnit = Unit.getSelectedDeliveryUnit(context)
		val tempUnit = TempUnit.getSelectedTemperatureUnit(context)
		val usingEc = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("tds_ec", false)

		var summary = ""
		var waterStr = StringBuilder()

		ph?.let {
			waterStr.append("<b>")
			waterStr.append(context.getString(R.string.plant_summary_ph))
			waterStr.append("</b>")
			waterStr.append(it)
			waterStr.append(", ")
		}

		runoff?.let {
			waterStr.append("<b>")
			waterStr.append(context.getString(R.string.plant_summary_out_ph))
			waterStr.append("</b>")
			waterStr.append(it)
			waterStr.append(", ")
		}

		summary += if (waterStr.toString().isNotEmpty()) waterStr.toString().substring(0, waterStr.length - 2) + "<br/>" else ""

		waterStr = StringBuilder()

		ppm?.let {
			var ppm = it.toLong().toString()
			if (usingEc)
			{
				waterStr.append("<b>EC: </b>")
				ppm = (it * 2.0 / 1000.0).toString()
			}
			else
			{
				waterStr.append("<b>PPM: </b>")
			}

			waterStr.append(ppm)
			waterStr.append(", ")
		}

		amount?.let {
			waterStr.append("<b>")
			waterStr.append(context.getString(R.string.plant_summary_amount))
			waterStr.append("</b>")
			waterStr.append(Unit.ML.to(deliveryUnit, it))
			waterStr.append(deliveryUnit.label)
			waterStr.append(", ")
		}

		temp?.let {
			waterStr.append("<b>")
			waterStr.append(context.getString(R.string.plant_summary_temp))
			waterStr.append("</b>")
			waterStr.append(TempUnit.CELCIUS.to(tempUnit, it))
			waterStr.append("º").append(tempUnit.label).append(", ")
		}

		summary += if (waterStr.toString().isNotEmpty()) waterStr.toString().substring(0, waterStr.length - 2) + "<br/>" else ""

		waterStr = StringBuilder()

		if (additives.size > 0)
		{
			waterStr.append("<b>")
			waterStr.append(context.getString(R.string.plant_summary_additives))
			waterStr.append("</b>")

			additives.forEach { additive ->
				if (additive.amount == null) return@forEach

				val converted = Unit.ML.to(measureUnit, additive.amount!!)
				val amountStr = if (converted == Math.floor(converted)) converted.toInt().toString() else converted.toString()

				waterStr.append("<br/>&nbsp;&nbsp;&nbsp;&nbsp;• ")
				waterStr.append(additive.description)
				waterStr.append("  -  ")
				waterStr.append(amountStr)
				waterStr.append(measureUnit.label)
				waterStr.append("/")
				waterStr.append(deliveryUnit.label)
			}
		}

		summary += waterStr.toString()

		return summary
	}
}
